### PR TITLE
RA-1621: Fix <encounterDiagnoses/> not submitting after <obsGroup/>.

### PR DESCRIPTION
**Pull request by Ruhanga**
…unterDiagnoses/> after empty values of <obsGroup/>

Ticket : https://issues.openmrs.org/browse/RA-1621

### CODE REVIEW

File: api-2.2/src/test/java/org/openmrs/module/htmlformentry/EncounterDiangnosesTagHandlerComponentTest.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    EmrApiProperties emrApiProperties;
+	
+	@Before
+	public void setup() throws Exception {
+        encounterDiagnosesTagHandler = CoreAppsActivator.setupEncounterDiagnosesTagHandler(conceptService, adtService, Context.getRegisteredComponent("emrApiProperties", EmrApiProperties.class), null);
+        Context.getService(HtmlFormEntryService.class).addHandler(CoreAppsConstants.HTMLFORMENTRY_ENCOUNTER_DIAGNOSES_TAG_NAME, encounterDiagnosesTagHandler);
+	}
+
+	@Test
+	public void testSingleObsGroupAndEncounterDiagnosesTagsShouldPassGivenNonEmptyObsGroup() throws Exception {
```
[the remark this answers is not in the document]
Comment (Ruhanga) [reply api-2.2/src/test/java/org/openmrs/module/htmlformentry/EncounterDiangnosesTagHandlerComponentTest.java]: Unfortunately, I reflexively clicked the full PR button for this, my bad :(
I'll definitely include both tests, Thanks.

File: api-2.2/pom.xml [hunk 2]
```diff
@@ -11,9 +11,20 @@
 	<name>Core Apps Module API 2.2</name>
 	<description>API project for CoreApps</description>
 	<properties>
```
Comment (mogoodrich) [inline api-2.2/pom.xml]: There seem to be a lot of changes to dependencies here... can you provide a summary of what you are changing and why?

File: api-2.2/src/test/java/org/openmrs/module/htmlformentry/EncounterDiangnosesTagHandlerComponentTest.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
+package org.openmrs.module.htmlformentry;
```
Comment (mogoodrich) [inline api-2.2/src/test/java/org/openmrs/module/htmlformentry/EncounterDiangnosesTagHandlerComponentTest.java]: Typo in file name ... Diagnoses not Diangnoses

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java [hunk 4]
```diff
@@ -329,8 +331,7 @@ public void handleSubmission(FormEntrySession formEntrySession, HttpServletReque
                     if (diagnosis.getEncounter().getEncounterType() == null) {
                         diagnosis.getEncounter().setEncounterType(formEntrySession.getForm().getEncounterType());
                     }
-
-                    Context.getEncounterService().saveEncounter(diagnosis.getEncounter());
+                    Context.getService(HtmlFormEntryService.class).applyActions(formEntrySession);
```
Comment (mogoodrich) [inline api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Ideally, we wouldn't be applying the actions directly from a component like this... the element would be creating new actions that would be executed later. 

But, yeah, that ship sailed a long time ago... there are many other examples of where directly modify stuff within a submission element, and this seems better than what we are currently doing... saving the encounter directly within the element.

File: api-2.2/src/test/java/org/openmrs/module/htmlformentry/EncounterDiangnosesTagHandlerComponentTest.java (hunk 3 above)
Comment (Ruhanga) [reply api-2.2/src/test/java/org/openmrs/module/htmlformentry/EncounterDiangnosesTagHandlerComponentTest.java]: Cool, thanks for this catch @mogoodrich.

File: api-2.2/pom.xml (hunk 2 above)
Comment (Ruhanga) [reply api-2.2/pom.xml]: Oh, yeah. Doing this. I brief, these dependencies are necessary for the context sensitive tests.

Comment (mks-d) [reply api-2.2/pom.xml]: @Ruhanga the they should be test-scoped:
```xml
<scope>test</scope>
```

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java (hunk 4 above)
Comment (mks-d) [reply api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: >the element would be creating new actions that would be executed later

@Ruhanga can you look into this?

File: api-2.2/pom.xml (hunk 2 above)
Comment (Ruhanga) [reply api-2.2/pom.xml]: Wouldn't the dependencies be needed for later usage by this sub-project @mks-d ?

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java (hunk 4 above)
Comment (Ruhanga) [reply api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: Yep, spiking on a better to handle the encounter diagnoses submission.

File: api-2.2/pom.xml (hunk 2 above)
Comment (mks-d) [reply api-2.2/pom.xml]: @Ruhanga until they are not needed outside of the tests they should remain test-scoped.

File: api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java (hunk 4 above)
Comment (Ruhanga) [reply api-2.2/src/main/java/org/openmrs/module/coreapps/htmlformentry/EncounterDiagnosesElement.java]: So what I've done here is avoid what was in the first place causing issues, empty obs groups, then save directly.

File: api-2.2/src/test/java/org/openmrs/module/htmlformentry/EncounterDiagnosesTagHandlerComponentTest.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    EmrApiProperties emrApiProperties;
+	
+	@Before
+	public void setup() throws Exception {
+        encounterDiagnosesTagHandler = CoreAppsActivator.setupEncounterDiagnosesTagHandler(conceptService, adtService, Context.getRegisteredComponent("emrApiProperties", EmrApiProperties.class), null);
+        Context.getService(HtmlFormEntryService.class).addHandler(CoreAppsConstants.HTMLFORMENTRY_ENCOUNTER_DIAGNOSES_TAG_NAME, encounterDiagnosesTagHandler);
+	}
+
+	@Test
+	public void testSingleObsGroupAndEncounterDiagnosesTagsShouldPassGivenNonEmptyObsGroup() throws Exception {
```
Comment (mks-d) [inline api-2.2/src/test/java/org/openmrs/module/htmlformentry/EncounterDiagnosesTagHandlerComponentTest.java]: → `testEncounterDiagnoses_shouldSubmitAfterNonEmptyObsGroup`

File: api-2.2/src/test/java/org/openmrs/module/htmlformentry/EncounterDiagnosesTagHandlerComponentTest.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+				results.assertProvider(1);
+				results.assertLocation(2);
+				results.assertObsCreated(51, "1x3 daily");
+				assertThat(Context.getDiagnosisService().getDiagnoses(getPatient(), date).get(0).getDiagnosis().getCoded().getId(), is(22));
+			}
+		}.run();
+	}
+	
+	@Test
+	public void testSingleObsGroupAndEncounterDiagnosesTagsShouldPassGivenEmptyObsGroup() throws Exception {
```
Comment (mks-d) [inline api-2.2/src/test/java/org/openmrs/module/htmlformentry/EncounterDiagnosesTagHandlerComponentTest.java]: → `testEncounterDiagnoses_shouldSubmitAfterEmptyObsGroup`

**Pull request merged**